Sources/NIOHTTP1/HTTPObjectAggregator.swift

```diff
@@ -0,0 +1,422 @@
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// The parts of a complete HTTP response from the view of the client.
+///
+/// Afull HTTP request is made up of a response header encoded by `.head`
+/// and an optional `.body`.
+public struct HTTPServerRequestFull {
+    public var head: HTTPRequestHead
+    public var body: ByteBuffer?
+
+    public init(head: HTTPRequestHead, body: ByteBuffer?) {
+        self.head = head
+        self.body = body
+    }
+}
+
+extension HTTPServerRequestFull: Equatable {}
+
+/// The parts of a complete HTTP response from the view of the client.
+///
+/// Afull HTTP response is made up of a response header encoded by `.head`
+/// and an optional `.body`.
+public struct HTTPClientResponseFull {
+    public var head: HTTPResponseHead
+    public var body: ByteBuffer?
+
+    public init(head: HTTPResponseHead, body: ByteBuffer?) {
+        self.head = head
+        self.body = body
+    }
+}
+
+extension HTTPClientResponseFull: Equatable {}
+
+public enum HTTPObjectAggregatorError: Error {
+    case frameTooLong
+    case connectionClosed
+    case messageDropped
+    case channelStateError(event: String, state: String)
+    case finishedIgnoringContent
+}
+
+extension HTTPObjectAggregatorError: Equatable {}
+
+public enum HTTPObjectAggregatorEvents {
+    case httpExpectationFailedEvent
+    case httpContinueEvent
+    case httpFrameTooLongEvent
+}
+
+/// The state of the aggregator  connection.
+internal enum AggregatorState {
+    /// Nothing is active on this connection, the next message we expect would be a request `.head`.
+    case idle
+
+    /// Ill-behaving client may be sending content that is too large
+    case ignoringContent
+
+    /// We are receiving and aggregating a request
+    case receiving
+
+    /// Connection should be closed
+    case closed
+
+    mutating func messageHeadReceived() throws {
+        switch self {
+        case .idle:
+            self = .receiving
+        case .ignoringContent, .receiving, .closed:
+            throw HTTPObjectAggregatorError.channelStateError(event: "head", state: "\(self)")
+        }
+    }
+
+    mutating func messageBodyReceived() throws {
+        switch self {
+        case .idle, .closed:
+            throw HTTPObjectAggregatorError.channelStateError(event: "body", state: "\(self)")
+        case .receiving, .ignoringContent:
+            ()
+        }
+    }
+
+
+    mutating func messageEndReceived() throws {
+        switch self {
+        case .receiving:
+            // Got the request end we were waiting for.
+            self = .idle
+        case .ignoringContent:
+            // Finished ignoring message contents
+            self = .idle
+            throw HTTPObjectAggregatorError.finishedIgnoringContent
```

Comment:
Is this a real error? What are we trying to do here?

Reply:
this would be an expected transition when an oversize message ends from essentially an "erroring" state where any content received is getting discarded. Throwing an error avoids going into dispatching the aggregated message and instead just cleaning up

Reply:
I'd love a comment to that effect here to explain why we are throwing an error here, and perhaps the error should be a named so as to be clear to the user who catches it. Right now I think its name favours the implementer, not the catcher.